keps/sig-cli/0015-kubectl-plugins.md

```diff
@@ -0,0 +1,151 @@
+  plugin that exists in an environment.
+* Plugin management is out of the scope of this design. A mechanism for updating and managing lifecycle of existing 
+  plugins should be covered as a separate design (See https://github.com/kubernetes/community/pull/2340).
+
+## Proposal
+
+### Scenarios
+
+* Developer wants to create and expose a plugin to `kubectl`.
+  They use a programming language of their choice and create an executable file.
+  The executable's filename consists of the command path to replace, and is prefixed with `kubectl-`.
+  The executable file is placed on the user's PATH, preferably under `/usr/local/bin`.
+  If a plugin is written using a dynamic language, and depends on logic defined in additional files, those
+  files are placed elsewhere in the user's system (e.g. ~/.kubeplugins/myplugin/helper1.py) and not in the user's `PATH`.
+* User wants to "install" a plugin `kubectl-plugin-foo`
+  They move the executable file `kubectl-plugin-foo` to a location on their path, and access the plugin via the command `kubectl plugin foo`.
+* User wants a subset of their plugin executables to live in a specific location on their filesystem.
+  The user creates a directory in the desired location (`~/Documents/foo-plugins`) and adds that location to their `PATH`.
+
+### Implementation Details/Notes/Constraints
+
+The current implementation details for the proposed design rely on using a plugin executable's name to determine what
+command the plugin is adding or overwriting.
+For a given command `kubectl plugin foo --bar baz`, an executable `kubectl-plugin-foo` will be matched on a user's `PATH`,
+and the arguments `--bar baz` will be passed to it in that order.
+
+A potential limitation of this could present itself in the order of arguments provided by a user.
+A user could intend to run a plugin `kubectl-plugin-foo-bar` with the flag `--baz` with the following command
+`kubectl foo --baz bar`, but instead end up matching `kubectl-foo` with the flag `--baz` and the argument `bar` based
+on the placement of the flag `--baz`.
+
+A notable constraint of this design is that it excludes any form of plugin lifesycle management, or version compatibility.
+A plugin may depend on other plugins based on the desicion of a plugin author, however the current design does nothing
+to facilitate such dependencies. It is up to the plugin's author (or a separate / independent plugin management system) to
+provide documentation or instructions on how to meet any dependencies required by a plugin.
+
+### Risks and Mitigations
+
+Certain command paths could be made significantly more complex to debug when plugins that overwrite them are taken into account.
+For example, `kubectl get foobar` could now be failing due to:
+- An actual bug with the `get` command
+- `foobar` might be a CRD that was not defined correctly
+- There could be a plugin overriding the path `kubectl-get-foobar` and that plugin is experiencing an error.
```

Comment:
is this design proposing allowing plugins to override every kubectl command/subcommand? is there a reason we wouldn't start with the existing plugin extension point and maybe the create command? letting a plugin override `kubectl get foobar` seems like a non-goal to me

Reply:
> is this design proposing allowing plugins to override every kubectl command/subcommand? 

No, at least it is not my intention for this to propose overriding any path. I meant to use this point as a means to justify the last paragraph in this section, which calls for a way to only allow plugins to override an allowed set of command paths.

Reply:
During one of the earlier sig-cli meetings we've agreed to start with not allowing to override the defaults. That will come as a natural evolution.